server/src/com/thoughtworks/go/server/web/BackupFilter.java

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright 2016 ThoughtWorks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.thoughtworks.go.server.web;
+
+import com.google.gson.JsonObject;
+import com.thoughtworks.go.server.service.BackupService;
+import org.apache.commons.io.IOUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.context.support.SpringBeanAutowiringSupport;
+import org.springframework.web.util.HtmlUtils;
+
+import javax.servlet.*;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.StringWriter;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * @understands redirecting all requests to a service unavailable page when the server is being backed up.
+ */
+public class BackupFilter implements Filter {
+
+    public static final String JSON = "json";
+    public static final String XML = "xml";
+
+    @Autowired
+    private BackupService backupService;
+    private final static org.apache.log4j.Logger LOGGER = org.apache.log4j.Logger.getLogger(BackupFilter.class);    // For the Test
+    public static final Pattern PATTERN = Pattern.compile("^((.*/api/.*)|(.*[^/]+\\.(xml|json)(\\?.*)?))$");
+
+    // For the Test
+    protected BackupFilter(BackupService backupService) {
+        this.backupService = backupService;
+    }
+
+    public BackupFilter() {
+
+    }
+
+    @Override
+    public void init(FilterConfig filterConfig) throws ServletException {
+    }
+
+    @Override
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
+        if (backupService == null) {
+            SpringBeanAutowiringSupport.processInjectionBasedOnCurrentContext(this);
+        }
+
+        String url = ((HttpServletRequest) request).getRequestURI();
+        ((HttpServletResponse) response).setHeader("Cache-Control", "private, max-age=0, no-cache");
+        ((HttpServletResponse) response).setDateHeader("Expires", 0);
```

Comment:
The cache control headers are to be inserted ONLY IF the filter is rendering the response, not if the filter is forwarding along the chain. You may end up setting cache headers on assets among other things significantly killing performance both for clients and the server.